### ISSUE TITLE
Support for multiple tagExpressions?

### ISSUE THREAD

**Issue by psovit** [jonsamwell/flutter_gherkin#57]
@jonsamwell 
Is there any way I can use multiple tagExpressions?
It would be easier if I could preload data differently based on different tagExpression.

**Comment by jonsamwell** [comment]
Hi @psovit,

Could you expand what you mean here?

Are you referring to something like this (which already works)

```
// see tagExpression: 'not @smoke and not @ignore and not @android'

{
            "name": "Debug Integration Tests (Local)",
            "program": "test_harness.dart",
            "cwd": "tests/integration",
            "request": "launch",
            "type": "dart",
            "args": [
                "--targetConfig",
                "local",
                "--buildFlavor",
                "development",
                "--logFlutterProcessOutput",
                "--buildTimeoutInSeconds",
                "360",
                "--tagExpression",
                "not @smoke and not @ignore and not @android",
            ],
        }
```

**Comment by psovit** [comment]
Hi @jonsamwell,

I need to run a script that preloads user data differently in server based upon different scenario. For this, right now I have hooks setup to run before scenario. So, what I needed to configure is something like this: 

if tagExpression is `@withProgress` run the script to load user progress,
if tagExpression is `@withBookmarks` run the script to load user bookmarks.

I was hoping if I can add these tags for different scenarios, the tests can dynamically preload data based on the tagExpression.

I hope I was clear for you to understand.

Thanks.

**Comment by jonsamwell** [comment]
@psovit so you would want to use a Hook and override say the `onBeforeScenario` method to setup the server data based on one or more tags on the scenario about to be tested.

Just so I am understanding this correctly:

If you have a scenario like this:
```
Feature: User's book library

  @scifi @satire
  Scenario: User's book library shows stored books
    Given I am logged in
    And I am on the book library page
   Then I see the books
   | Title                                | Genre  |
   | Catch 22                         | satire   |
   | Rendezvous with Rama  | scifi     |
```

based on the above tags you would setup some data on the server that the app serves up in the screens during the test?

```
class DataPreLoadingHook extends Hook {
 Future<void> onBeforeScenario(TestConfiguration config, String scenario) async {
     // need some way of getting the scenario tags?
    // then you could do the data pre-loading
    await api.preload(Genres.scifi);
  }
}
```

Would that be correct?

On a side note are you sure this is a complex anitpattern as your test is not really explaining the correct scenario.  I always do the way pre-loading for the server using the `Background` block, for example:

```
Feature: User's book library
  Background: Pre-load account data
   Given a new user account
   And their book store is populated with the books
   | Title                               | Genre  |
   | Catch 22                        | satire   |
   | Rendezvous with Rama | scifi     |

  @scifi @satire
  Scenario: User's book library shows stored books
    Given I am logged in
    And I am on the book library page
   Then I see the books
   | Title                                | Genre  |
   | Catch 22                         | satire   |
   | Rendezvous with Rama  | scifi     |
```

**Comment by psovit** [comment]
I had never made use of the `Background` block and didn't know it existed :) Thanks I think if I use this block I can definitely write re-usable steps here. One question on it's usage though, can I use multiple `Background` within the same feature file like grouping scenarios by the background block?

Coming back to my current implementation, yes you are right that is how I am making use of the onBeforeScenario hook.

However, one thing is in my feature files I have only made use of single tagExpression `@testable`. So, in the `app_test` configuration I have `..tagExpression = "@testable"`.

This allows me to run only those scenarios which have been implemented and ignore the ones which are yet to be implemented from that feature file.

So, now I was trying to see if I can use another tagExpression along with `@testable` so that I can achieve running only testable scenarios as well as preloading data based on different tag use.

**Comment by jonsamwell** [comment]
Unfortunately you can only have a single `Background` block per feature file.  I only put a single scenario per file anyway to reduce the file complexity and keep my tests simple.

Yes, going back to your original requirement - I think as there is no way for a `Hook` to get the tags a certain `Scenario` block has it would be tough to do at the moment.  I can make a change to pass in the tags scenario focused hook methods and you should be able to achieve this.

On another not I usually add an `@ignore` tag to tests that are not complete so I can have a tag expression like this `@smoke and not @ignore`

**Comment by psovit** [comment]
When you say "I can make a change to pass in the tags scenario focused hook methods and you should be able to achieve this.", do you mean in the next version of the plugin?

About adding the `@ignore` tag, since we had a lot of scenarios in the beginning, we chose to identify only those that are testable rather than tagging all as ignoreable.

**Comment by jonsamwell** [comment]
Yes, sorry - it'll need plugin code update - I should have a bit of time tomorrow to sort this as it isn't too big of a change

**Comment by psovit** [comment]
That sounds wonderful @jonsamwell .
Again I highly appreciate your effort and prompt responses.

Have a great day!

**Comment by jonsamwell** [comment]
@psovit should be fixed in the latest version (1.1.7+6)! See the below link for an example

https://github.com/jonsamwell/flutter_gherkin/blob/master/example/test_driver/hooks/hook_example.dart#L28

 Let me know how you get on - closing this for now.
